pkg/datapath/orchestrator/orchestrator.go

```diff
@@ -39,46 +72,239 @@ type orchestratorParams struct {
 	DirectRoutingDevice tables.DirectRoutingDevice
 	LocalNodeStore      *node.LocalNodeStore
 	NodeDiscovery       *nodediscovery.NodeDiscovery
+	JobRegistry         job.Registry
+	Health              cell.Health
+	Lifecycle           cell.Lifecycle
+	EndpointManager     endpointmanager.EndpointManager
+	ConfigPromise       promise.Promise[*option.DaemonConfig]
 }
 
 func newOrchestrator(params orchestratorParams) *orchestrator {
-	return &orchestrator{
-		params: params,
+	o := &orchestrator{
+		params:        params,
+		trigger:       make(chan reinitializeRequest, 1),
+		dpInitialized: make(chan struct{}),
 	}
-}
 
-func (o *orchestrator) Reinitialize(ctx context.Context) error {
-	// Wait until the local node has been populated by NodeDiscovery.
-	o.params.NodeDiscovery.WaitForLocalNodeInit()
+	group := params.JobRegistry.NewGroup(params.Health)
+	group.Add(job.OneShot("Reinitialize", o.reconciler, job.WithShutdown()))
+	params.Lifecycle.Append(group)
 
-	localNode, err := o.params.LocalNodeStore.Get(ctx)
-	if err != nil {
-		return fmt.Errorf("get local node: %w", err)
-	}
-
-	rxn := o.params.DB.ReadTxn()
-	directRoutingDevice, _ := o.params.DirectRoutingDevice.Get(ctx, rxn)
+	return o
+}
 
-	// Construct the LocalNodeConfiguration that encapsulates the
-	// local node's dynamic configuration.
-	localNodeConfig, err := newLocalNodeConfig(
+func (o *orchestrator) reconciler(ctx context.Context, health cell.Health) error {
+	localNodeConfig, _, _, _, err := newLocalNodeConfig(
+		ctx,
 		option.Config,
-		localNode,
+		node.LocalNode{},
 		o.params.MTU,
-		rxn,
+		o.params.DB.ReadTxn(),
+		o.params.DirectRoutingDevice,
 		o.params.Devices,
-		directRoutingDevice,
 		o.params.NodeAddresses,
 	)
 	if err != nil {
-		return fmt.Errorf("build LocalNodeConfiguration: %w", err)
+		health.Degraded("failed to get local node configuration", err)
+	}
+
+	// We have to reinitialize the host device before we wait for the daemon config.
+	// That is because the daemon startup expects the host device to be initialized
+	// before it will populate the config with the values we are looking for.
+	if err := o.params.Loader.ReinitializeHostDev(ctx, &localNodeConfig); err != nil {
```

Comment:
There should be a wait on the daemon startup side for this device to appear.

Reply:
Do wonder if we could move those things here somewhere. daemon/cmd directly accessing the host device seems very wrong.